### PR TITLE
Crossgen2: ensure base address >= 4 GiB for 64-bit Windows targets

### PR DESCRIPTION
**Pull request by AntonLapounov**
Base addresses below 4 GiB are reserved for WoW on x64 and disallowed on ARM64.  If the input assembly was compiled for anycpu, its base address is 32-bit and we need to fix it.

Also fix ADRP's argument calculation for ARM64.

### CODE REVIEW

File: src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs [hunk 1]
```diff
@@ -118,7 +118,7 @@ public ModuleDesc GetModuleForSimpleName(string simpleName, bool throwIfNotFound
                 if (!ReferenceFilePaths.TryGetValue(simpleName, out filePath))
                 {
                     // We allow the CanonTypesModule to not be an EcmaModule.
-                    if (((IAssemblyDesc)CanonTypesModule).GetName().Name == simpleName)
+                    if (CanonTypesModule != null && ((IAssemblyDesc)CanonTypesModule).GetName().Name == simpleName)
```
Comment (jkotas) [inline src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs]: Should we rather throw earlier when there is no system module?

Something like https://github.com/dotnet/corert/pull/7972 ?

cc @MichalStrehovsky

Comment (AntonLapounov) [reply src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs]: Great, that PR addresses the issue, so I will revert this part.

Comment (AntonLapounov) [reply src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs]: Oh, that is a CoreRT fix. I may add the same code for crossgen2.

**Review by jkotas — APPROVED** [review]

File: src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs (hunk 1 above)
Comment (AntonLapounov) [reply src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs]: Reverted this change in favor of a better fix — need to port dotnet/corert#7972.

Comment (MichalStrehovsky) [reply src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs]: Yep, checking earlier is better than trying to make the system resilient to bad state.

I never liked the `SetSystemModule` API. It's a wart. Type system context shouldn't have mutable state that makes it possible to use it before it's ready to be used.

**Pull request merged**